crypto/fipsmodule/hmac/hmac.c

```diff
@@ -393,6 +477,125 @@ void HMAC_CTX_reset(HMAC_CTX *ctx) {
+  *out_len = chaining_length * 2;
+  assert(*out_len <= HMAC_MAX_PRECOMPUTED_KEY_SIZE);
+  if (NULL == out) {
+    // When out is NULL, we just set out_len.
+    // We keep the state as HMAC_STATE_PRECOMPUTED_KEY_EXPORT_READY
+    // to allow an actual export of the precomputed key immediately afterward.
+    return 1;
+  }
+  uint64_t i_ctx_n;
+  uint64_t o_ctx_n;
+
+  const int ok = ctx->methods->get_state(&ctx->i_ctx, out, &i_ctx_n) &&
+      ctx->methods->get_state(&ctx->o_ctx, out + chaining_length, &o_ctx_n);
+
+  // We should never arrive here as in our setting, get_state should always be
+  // successful since i_ctx/o_ctx have processed exactly one block
+  assert(ok);
+
+  // Sanity check: we must have processed a single block at this time
+  size_t block_size = EVP_MD_block_size(ctx->md);
+  assert(8 * block_size == i_ctx_n);
+  assert(8 * block_size == o_ctx_n);
+
+  // The context is ready to be used to compute HMAC values at this point.
+  ctx->state = HMAC_STATE_INIT_NO_DATA;
+
+  return 1;
+}
+
+int HMAC_Init_from_precomputed_key(HMAC_CTX *ctx,
+                                   const uint8_t *precomputed_key,
+                                   size_t precomputed_key_len,
+                                   const EVP_MD *md) {
+  if (HMAC_STATE_READY_NEEDS_INIT == ctx->state ||
+      HMAC_STATE_PRECOMPUTED_KEY_EXPORT_READY == ctx->state) {
+    ctx->state = HMAC_STATE_INIT_NO_DATA;  // Mark that init has been called
+  }
+
+  if (HMAC_STATE_INIT_NO_DATA == ctx->state) {
+    if (precomputed_key == NULL && (md == NULL || md == ctx->md)) {
+      // If nothing is changing then we can return without doing any further
+      // work.
+      return 1;
+    }
+  }
+
+  // Now we assume that we need to re-initialize everything.
+  // See HMAC_Init_ex
+
+  if (!hmac_ctx_set_md_methods(ctx, md)) {
+    return 0;
+  }
+
+  const HmacMethods *methods = ctx->methods;
+
+  const size_t chaining_length = methods->chaining_length;
+  const size_t block_size = EVP_MD_block_size(methods->evp_md);
+  assert(block_size <= EVP_MAX_MD_BLOCK_SIZE);
+  assert(2 * chaining_length <= HMAC_MAX_PRECOMPUTED_KEY_SIZE);
+
+  if (2 * chaining_length != precomputed_key_len) {
+    return 0;
+  }
+
+  // We require precomputed_key to be non-NULL, since here md changed
+  if (NULL == precomputed_key) {
+    return 0;
+  }
+
+  // We have to avoid the underlying SHA services updating the indicator
+  // state, so we lock the state here. Technically this is not really needed,
+  // because the functions we call should not update the indicator state.
+  // But this is safer.
+  FIPS_service_indicator_lock_state();
+  int result = 0;
+
+  // Initialize i_ctx from the state stored in the first part of precomputed_key
+  // Recall that i_ctx is the state of the hash function after processing
+  // one block (ipad xor keyOrHashedKey)
+  if (!methods->init_from_state(&ctx->i_ctx, precomputed_key, block_size * 8)) {
```

Comment:
Do we need to zero out `i_ctx` and `o_ctx` first? Or maybe zero out the `data` and `num` members of them in the implementation of `init_from_state`?

Reply:
Yes, `init_from_state` starts by zeroing out everything:
https://github.com/aws/aws-lc/pull/1574/files#diff-45faa6f87b7148d297f13d80c973233a2dbfc4670d75621f402f729a9b109981R110

Reply:
Sorry, missed that, thank you.